external/storm-hdfs/src/main/java/org/apache/storm/hdfs/spout/HdfsSpout.java

```diff
@@ -0,0 +1,654 @@
+  private final Timer commitTimer = new Timer();
+  private boolean fileReadCompletely = false;
+
+  private String configKey = Configs.DEFAULT_HDFS_CONFIG_KEY; // key for hdfs kerberos configs
+
+  public HdfsSpout() {
+  }
+
+  public Path getLockDirPath() {
+    return lockDirPath;
+  }
+
+  public SpoutOutputCollector getCollector() {
+    return collector;
+  }
+
+  public HdfsSpout withConfigKey(String configKey){
+    this.configKey = configKey;
+    return this;
+  }
+
+  public void nextTuple() {
+    LOG.debug("Next Tuple");
+    // 1) First re-emit any previously failed tuples (from retryList)
+    if (!retryList.isEmpty()) {
+      LOG.debug("Sending from retry list");
+      HdfsUtils.Pair<MessageId, List<Object>> pair = retryList.remove();
+      emitData(pair.getValue(), pair.getKey());
+      return;
+    }
+
+    if( ackEnabled  &&  tracker.size()>=maxDuplicates ) {
+      LOG.warn("Waiting for more ACKs before generating new tuples. " +
+               "Progress tracker size has reached limit {}"
+              , maxDuplicates);
+      // Don't emit anything .. allow configured spout wait strategy to kick in
+      return;
+    }
+
+    // 2) If no failed tuples, then send tuples from hdfs
+    while (true) {
+      try {
+        // 3) Select a new file if one is not open already
+        if (reader == null) {
+          reader = pickNextFile();
+          if (reader == null) {
+            LOG.info("Currently no new files to process under : " + sourceDirPath);
+            return;
+          }
+        }
+
+        // 4) Read record from file, emit to collector and record progress
+        List<Object> tuple = reader.next();
+        if (tuple != null) {
+          fileReadCompletely= false;
+          ++tupleCounter;
+          MessageId msgId = new MessageId(tupleCounter, reader.getFilePath(), reader.getFileOffset());
+          emitData(tuple, msgId);
+
+          if(!ackEnabled) {
+            ++acksSinceLastCommit; // assume message is immediately acked in non-ack mode
+            commitProgress(reader.getFileOffset());
+          } else {
+            commitProgress(tracker.getCommitPosition());
+          }
+          return;
+        } else {
+          fileReadCompletely = true;
+          if(!ackEnabled) {
+            markFileAsDone(reader.getFilePath());
+          }
+        }
+      } catch (IOException e) {
+        LOG.error("I/O Error processing at file location " + getFileProgress(reader), e);
+        // don't emit anything .. allow configured spout wait strategy to kick in
+        return;
+      } catch (ParseException e) {
+        LOG.error("Parsing error when processing at file location " + getFileProgress(reader) +
+                ". Skipping remainder of file.", e);
+        markFileAsBad(reader.getFilePath());
+        // note: Unfortunately not emitting anything here due to parse error
+        // will trigger the configured spout wait strategy which is unnecessary
```

Comment:
@roshannaik 
Quick question: Is it safe to let reader as it is? Do we want to make reader as null to trigger finding next file?